substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/substitute/UnsafeAutomaticSubstitutionProcessor.java

```diff
@@ -682,6 +702,26 @@ private static ResolvedJavaField extractValueStoreField(ValueNode valueNode, Kin
             if (signExtendFieldStoreUsages.count() == 1) {
                 offsetField = signExtendFieldStoreUsages.first().field();
             }
+        } else {
+            // Determine if the usage of the offset escapes from local usage.
+            for (Node valueNodeUsage : valueNodeUsages) {
+                // Probably other valid node types could be added.
+                if (valueNodeUsage instanceof FrameState || valueNodeUsage instanceof MethodCallTargetNode) {
+                    continue;
+                }
+                if (valueNodeUsage instanceof ValuePhiNode) {
+                    NodeIterable<StoreFieldNode> phiStores = valueNodeUsage.usages().filter(StoreFieldNode.class);
+                    // Store is on the far side of a Phi node, probably due to a temporary variable
+                    // being used involving a try/catch and a final field which desires single
+                    // assignment.
+                    if (phiStores.count() == 1) {
+                        offsetField = phiStores.first().field();
```

Comment:
So here you follow a single (out of possibly many) usages. That is not sound, since one of the other usages can actually be a disallowed usage. Or a store to another field that we are then missing.

Also the phi function has other inputs, i.e., the offset is merged with a different value. This means that the `offsetField` might be initialized with a different (non-field-offset) value in some cases.

In summary, I don't think there is a good way to handle phi nodes, that's why we hadn't added such a handling yet.

Reply:
Yeah... I was wondering. I'm still learning my way around Phi nodes and such. I'll revert this portion.